### ISSUE TITLE
expertflyer detect bot

### ISSUE THREAD

**Issue by shimuldn** [berstend/puppeteer-extra#314]
https://www.expertflyer.com detect the puppeteer-extra and block the ip address sometimes sometimes show captcha but block most of times. There are using https://www.distilnetworks.com service.

puppeteer-extra headless false working fine.

**Comment by shimuldn** [comment]
> How to reproduce it? I've checked and do not see any blocks. Provide your code.




```

//sudo npm install puppeteer puppeteer-extra puppeteer-extra-plugin-stealth puppeteer-extra-plugin-adblocker readline
var headless_mode = process.argv[2]

const readline = require('readline');
const puppeteer = require('puppeteer-extra')
const StealthPlugin = require('puppeteer-extra-plugin-stealth')
puppeteer.use(StealthPlugin())
const AdblockerPlugin = require('puppeteer-extra-plugin-adblocker')
puppeteer.use(AdblockerPlugin({ blockTrackers: true }))


async function run () {
  const browser = await puppeteer.launch({
    headless:(headless_mode !== 'true')? false : true,
    ignoreHTTPSErrors: true,
    slowMo: 0,
    args: ['--window-size=1380,800',
    '--remote-debugging-port=9222',
    "--remote-debugging-address=0.0.0.0",
    '--disable-gpu', "--disable-features=IsolateOrigins,site-per-process", '--blink-settings=imagesEnabled=true'
    ]})

  const page = await browser.newPage();


  console.log(`Testing expertflyer.com`)
  //await page.goto('https://www.expertflyer.com')
  await goto_Page('https://www.expertflyer.com')
  await waitForNetworkIdle(page, 3000, 0)
  //await page.waitFor(7000)
  await checking_error(do_2nd_part)




  async function do_2nd_part(){
    try{await page.click('#yui-gen2 > a')}catch{}
    await page.waitFor(5000)
    var seat = '#headerTitleContainer > h1'
    try{console.log(await page.$eval(seat, e => e.innerText))}catch{}
    await page.screenshot({ path: 'expertflyer1.png'})

    await checking_error(do_3nd_part)
  }

  async function do_3nd_part(){
    try{await page.click('#yui-gen1 > a')}catch{}
    await page.waitFor(5000)
    var pro = '#headerTitleContainer > h1'
    try{console.log(await page.$eval(pro, e => e.innerText))}catch{}
    await page.screenshot({ path: 'expertflyer2.png'})

    console.log(`All done, check the screenshots?`)
  }


  async function checking_error(callback){
    try{
      try{var error_found = await page.evaluate(() => document.querySelectorAll('a[class="text yuimenubaritemlabel"]').length)}catch(error){console.log(`catch error ${error}`)}

      if (error_found === 0) {
        console.log(`Error found`)
        var captcha_msg = "Due to suspicious activity from your computer, we have blocked your access to ExpertFlyer. After completing the CAPTCHA below, you will immediately regain access unless further suspicious behavior is detected."
        var ip_blocked = "Due to recent suspicious activity from your computer, we have blocked your access to ExpertFlyer. If you feel this block is in error, please contact us using the form below."
        try{var error_msg = await page.$eval('h2', e => e.innerText)}catch{}
        try{var error_msg_details = await page.$eval('body > p:nth-child(2)', e => e.innerText)}catch{}

        if (error_msg_details == captcha_msg) {
          console.log(`Google Captcha found, You have to solve the captch here manually or some automation recaptcha service`)

          await verify_User_answer()
          await callback()
        } else if (error_msg_details == ip_blocked) {
          console.log(`The current ip address is blocked. The only way is change the ip address.`)
        } else {
          console.log(`Waiting for error page load... Waiting for 10 sec before rechecking...`)
          await page.waitFor(10000)
          await checking_error()
        }

      } else {
        console.log(`Page loaded successfully! You can do things here.`)
        await callback()
      }

    }catch{}
  }

  async function goto_Page(page_URL){
    try{
      await page.goto(page_URL, { waitUntil: 'networkidle2', timeout: 30000 });
    } catch {
      console.log(`Error in loading page, re-trying...`)
      await goto_Page(page_URL)
    }
  }

  async function verify_User_answer(call_back){
      user_Answer = await readLine();

      if (user_Answer == 'yes') {
        console.log(`user_Answer is ${user_Answer}, Processing...`)
        // Not working what i want. Will fix later
        // Have to restart the bot after solving
        await call_back()
      } else {
        console.log(`answer not match. try again...`)

        var user_Answer = await readLine();
        console.log(`user_Answer is ${user_Answer}`)
        await verify_User_answer(call_back)
      }
    }

    async function readLine() {

      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
      });

      return new Promise(resolve => {

        rl.question('Solve the captch and type yes to continue: ', (answer) => {
          rl.close();
          resolve(answer)
        });
      })
    }

  async function waitForNetworkIdle(page, timeout, maxInflightRequests = 0) {
  console.log('waitForNetworkIdle called')
  page.on('request', onRequestStarted);
  page.on('requestfinished', onRequestFinished);
  page.on('requestfailed', onRequestFinished);

  let inflight = 0;
  let fulfill;
  let promise = new Promise(x => fulfill = x);
  let timeoutId = setTimeout(onTimeoutDone, timeout);
  return promise;

  function onTimeoutDone() {
    page.removeListener('request', onRequestStarted);
    page.removeListener('requestfinished', onRequestFinished);
    page.removeListener('requestfailed', onRequestFinished);
    fulfill();
  }

  function onRequestStarted() {
    ++inflight;
    if (inflight > maxInflightRequests)
      clearTimeout(timeoutId);
  }

  function onRequestFinished() {
    if (inflight === 0)
      return;
    --inflight;
    if (inflight === maxInflightRequests)
      timeoutId = setTimeout(onTimeoutDone, timeout);
  }
}


  await browser.close()
}
run();


```

Getting random captcha but working.

I was playing with this code. If i disable `puppeteer.use(StealthPlugin())`  for a single run the ip got blocked and after 

enabling `puppeteer.use(StealthPlugin())`
`headless: false` working on the same blocked ip.
`headless: true` NOT working on the same blocked ip.

disabling `//puppeteer.use(StealthPlugin())`
`headless: false` NOT working on the same blocked ip.

**Comment by trnj** [comment]
I think  1. you do not use a proper user agent  2. window-size=1380,800 is a bit rare screen resolution

**Comment by trnj** [comment]
Try here  https://pixelscan.net/   your user agent should be the same OS as the platform.

**Comment by shimuldn** [comment]
> I think 1. you do not use a proper user agent 2. window-size=1380,800 is a bit rare screen resolution

So 1440,900 will do that.

Can you please help me with the user agent?

puppeteer-extra-plugin-anonymize-ua not in this case right?

**Comment by trnj** [comment]
Send a screenshot of https://pixelscan.net/

**Comment by shimuldn** [comment]
> Send a screenshot of https://pixelscan.net/

https://prnt.sc/uf4shc

**Comment by shimuldn** [comment]
Fixed https://prnt.sc/uf4vih

**Comment by trnj** [comment]
The good news: **puppeteer-extra works.**

I've tried this site on my setup and never got blocked. 

Then I used the part of your code purely and got blocked after 3-4 tries. So something in your code triggers distil.

Possible reasons I can figure:
- Upgrade the version of this plugin
- Use Win32 user agent and platform
- Start maximized 

Distil blocks for 1-2 hours (+/-) so you need to change IP or wait.

**Comment by shimuldn** [comment]
> The good news: **puppeteer-extra works.**
> 
> I've tried this site on my setup and never got blocked.
> 
> Then I used the part of your code purely and got blocked after 3-4 tries. So something in your code triggers distil.
> 
> Possible reasons I can figure:
> 
> * Upgrade the version of this plugin
> * Use Win32 user agent and platform
> * Start maximized
> 
> Distil blocks for 1-2 hours (+/-) so you need to change IP or wait.

Thats great.
Mine are updated.

````
const uaWin = UserAgentOverride({
  userAgent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.102 Safari/537.36",
  locale: 'en-US,en',
  platform: 'Win32'
})
````
Still masking fingerprint error. What i am doing wrong?

**Comment by trnj** [comment]
Latest version sets Win32 by default. 

Also try
--start-maximized
--window-size=1600,900

**Comment by shimuldn** [comment]
Still getting captcha

**Comment by trnj** [comment]
Yes, I was able to reproduce it on dedicated services. 

Even on my proven setup, but from a hosting network, I got distil error from https://www.expertflyer.com/ 

Probably a bug in puppeteer-extra

**Comment by evading-bot-detection** [comment]
I was able to access this just fine with the standard stealth package. Perhaps your IP is getting flagged?

**Comment by trnj** [comment]
> I was able to access this just fine with the standard stealth package. Perhaps your IP is getting flagged?

No, it not the issue. I've also reproduced it on the home network.

**Comment by trnj** [comment]
**I've found the issue**

It is the same I reported yesterday and then closed 
https://github.com/berstend/puppeteer-extra/issues/310

I downgraded from 2.6.1 to 2.6.0 and it works again. 

So the problem is in the 2.6.1.

**Comment by shimuldn** [comment]
> **I've found the issue**
> 
> It is the same as I reported yesterday and then closed
> #310
> 
> I downgraded from 2.6.1 to 2.6.0 and it works again.
> 
> So the problem is in the 2.6.1.

Yes, i just verified the same issue.

**Comment by Granitosaurus** [comment]
@shimuldn I'm looking through [2.6.1 and 2.6.0 diff](https://github.com/berstend/puppeteer-extra/compare/puppeteer-extra-plugin-stealth@2.6.0...puppeteer-extra-plugin-stealth@2.6.1)

Seems like the only evasion strategy change is navigator.plugins one unless there's a bug with `util` helper introduction. There seems to be minor but very important PR for plugins evasion: https://github.com/berstend/puppeteer-extra/pull/312 Could you try if this resolves 
Could you see if it solves this?

**Comment by trnj** [comment]
Yep, waiting for the fix. 

Btw, as I can see, you use Oracle Cloud. Does distil block IPs from this network by default (with 2.6.0)? OVH, hetzner, leaseweb, etc. are blocked by distil based on IP origin. Do they block Oracle network the same way?

**Comment by shimuldn** [comment]
> Yep, waiting for the fix.
> 
> Btw, as I can see, you use Oracle Cloud. Does distil block IPs from this network by default (with 2.6.0)? OVH, hetzner, leaseweb, etc. are blocked in distil based on IP origin, so not suitable for crawling. Do they block Oracle network the same way?

I don't do much crawling. I was testing on OC and others before you told me 2.6.0 to use.
Most of time i use gcp https://prnt.sc/ufy218
expertflyer.com is working without any issue on 2.6.0

**Comment by shimuldn** [comment]
I was able to pass the fingerprint for using Linux userAgent on Linux.
Are you able to pass the Win and Mac fingerprint on Linux?

**Comment by Niek** [comment]
Please check with this PR: https://github.com/berstend/puppeteer-extra/pull/341

**Comment by berstend** [comment]
Potential fix published in `puppeteer-extra-plugin-stealth@2.6.2`

**Comment by shimuldn** [comment]
> puppeteer-extra-plugin-stealth@2.6.2

Yes, solved.
Thank you.
